tests/cross/src/test/scala/tests/pc/InsertInferredTypeSuite.scala

```diff
@@ -511,6 +511,53 @@ class InsertInferredTypeSuite extends BaseCodeActionSuite {
        |""".stripMargin
   )
 
+  checkEdit(
+    "backticks-4".tag(IgnoreScala3),
+    """|case class `Foo-Foo`(i: Int)
+       |object O{
+       |  val <<foo>> = `Foo-Foo`(1)
+       |}""".stripMargin,
+    """|case class `Foo-Foo`(i: Int)
+       |object O{
+       |  val foo: `Foo-Foo` = `Foo-Foo`(1)
+       |}
+       |""".stripMargin
+  )
+
+  checkEdit(
+    "backticks-5".tag(IgnoreScala3),
+    """|object A{
+       |  case class `Foo-Foo`(i: Int)
+       |}
+       |object O{
+       |  val <<foo>> = A.`Foo-Foo`(1)
+       |}""".stripMargin,
+    """|object A{
+       |  case class `Foo-Foo`(i: Int)
+       |}
+       |object O{
+       |  val foo: A.`Foo-Foo` = A.`Foo-Foo`(1)
+       |}
+       |""".stripMargin
+  )
+
+  checkEdit(
+    "backticks-6".tag(IgnoreScala3),
+    """|object A{
+       |  case class `Foo-Foo`[A](i: A)
+       |}
+       |object O{
+       |  val <<foo>> = A.`Foo-Foo`(1)
+       |}""".stripMargin,
+    """|object A{
+       |  case class `Foo-Foo`[A](i: A)
+       |}
+       |object O{
+       |  val foo: A.`Foo-Foo`[Int] = A.`Foo-Foo`(1)
+       |}
+       |""".stripMargin
+  )
```

Comment:
test case that doesn't work with current approach
```scala
  checkEdit(
    "backticks-7".tag(IgnoreScala3).only,
    """|object A{
       |  class `x-x`()
       |  case class Foo[A](i: A)
       |}
       |object O{
       |  val <<foo>> = A.Foo(new A.`x-x`())
       |}""".stripMargin,
    """|object A{
       |  class `x-x`
       |  case class Foo[A](i: A)
       |}
       |object O{
       |  val foo: A.Foo[`x-x`] = A.Foo(new A.`x-x`)
       |}
       |""".stripMargin
  )
```

Reply:
I started looking at this PR and it's lost on what is wrong with the current approach. I played around with this a bit and seems to work fine for me.
This test also passes:
```
  checkEdit(
    "backticks-7".tag(IgnoreScala3).only,
    """|object A{
       |  class `x-x`
       |  case class Foo[A](i: A)
       |}
       |object O{
       |  val <<foo>> = A.Foo(new A.`x-x`)
       |}""".stripMargin,
    """|object A{
       |  class `x-x`
       |  case class Foo[A](i: A)
       |}
       |object O{
       |  val foo: A.Foo[A.`x-x`] = A.Foo(new A.`x-x`)
       |}
       |""".stripMargin
  )
  ```
Do you have any case when `sym.decodedName` is different than the string?
What type was inferred in this test for you?

Reply:
hmm. Honestly, the only I remember is that I got curious by these clauses - https://github.com/scalameta/metals/blob/c1d7984b5e8de705177743182bf5c714a01abbb2/mtags/src/main/scala-2/scala/meta/internal/pc/MetalsGlobal.scala#L190-L193

Maybe tests might spot some issues in approach if you remove them

If you won't find anything, feel free to merge this PR